recipes/lora_ppo_single_device.py

```diff
@@ -0,0 +1,813 @@
+                        ).bool()
+
+                    # run reward model on truncated query-response sequences: shape [b, context_length + max_generated_tokens]
+                    # TODO (SalmanMohammadi): Add support for _reward_model and _model using different tokenizers
+                    scores = self._reward_model(
+                        query_responses, input_pos=position_ids, mask=masks
+                    )
+
+                    idxs = utils.get_last_non_masked_token(padding_masks)
+                    # shape [b, ]
+                    scores = scores[
+                        torch.arange(0, scores.shape[0]), idxs + context_length
+                    ].squeeze(-1)
+
+                    reward_penality_mask = torch.zeros_like(scores).to(bool)
+
+                    # to reviewer - see commment above
+                    # - sequences without a EOS ID recieve a score of -1
+                    if self.penalise_no_eos:
+                        reward_penality_mask = ~eos_mask.any(-1)
+                    # - sequences with length < truncate_after_tokens recieve a score of -1
+                    if self.truncate_after_tokens is not None:
+                        reward_penality_mask |= ~(
+                            eos_mask.cumsum(-1).sum(-1) - 1
+                            <= self.truncate_after_tokens
+                        )
+
+                    # see https://arxiv.org/pdf/1909.08593 section 3.1.2
+                    scores.masked_fill_(reward_penality_mask, self.reward_penalty)
+
+                    # now mask out logprobs and values w.r.t. padding masks
+                    # https://github.com/huggingface/trl/blob/
+                    #   f5168fdbaf9cbf6a3f1bdc64dc44b9db3a9ae333/trl/trainer/ppov2_trainer.py#L359
+                    logprobs = logprobs.masked_fill(padding_masks, 1.0)
+                    ref_logprobs = ref_logprobs.masked_fill(padding_masks, 1.0)
+
+                    # values are masked up to and including the first padding token
+                    # see https://github.com/huggingface/trl/blob/f5168fdbaf9cbf6a3f1bdc64dc44b9db3a9ae333/
+                    #   trl/trainer/ppov2_trainer.py#L354
+                    # and the link to the excalidaw therein for a visual explanation
+                    value_padding_masks = padding_masks.clone()
+                    valid_response_lengths = (
+                        self.max_generated_tokens
+                        - value_padding_masks.cumsum(dim=-1)[:, -1]
+                    )
+                    valid_response_lengths = valid_response_lengths.clip(
+                        0, self.max_generated_tokens - 1
+                    )
+                    value_padding_masks[
+                        torch.arange(input_ids.shape[0]), valid_response_lengths
+                    ] = False
+                    values = values[:, context_length - 1 : -1].squeeze(
+                        -1
+                    )  # [b, max_generated_tokens]
```

Comment:
Sorry I'm not sure I fully follow this block. What exactly are we aiming to mask here? Also, I see the `context_length - 1 : -1` slice coming up quite a bit below as well, is this just the last padding token plus all the generated tokens, stopping just short of the eos token?

Reply:
The `context_length - 1 : -1` doesn't actually have anything to do with padding/eos tokens, it's because we're always predicting the next token, so the logprobs are a step behind. Consider that at each step in the response, the action taken is `a_t = response_t`, and the subsequent state produced by the "environment",  is `query_0, ... query_n, response_0, ...response_t` (i.e. the environment dynamics are just whichever sampling strategy you choose).

The `context_length - 1 : -1` happens because we're estimating logprobs of our policy for every action `pi_theta(a_t, s_t)`. So at the start of the trajectory, the logprobs for the first response token, `pi_theta(response_0 | query_0, ..., query_n)` are `F.log_softmax(self._policy_model([query_0, ..., query_n])`, hence why we do `context_length - 1`. Similarly, we slice up to `:-1` because the logprobs for the last action are at position -1, and there are no more actions taken (tokens generated) to estimate logprobs for.